### ISSUE TITLE
Process cache is invalidated by unexpected input file timestamp change

### ISSUE THREAD

**Issue by fmorency** [nextflow-io/nextflow#828]
### Expected behavior and actual behavior

I've had some issues with processes that are re-executed using `-resume` when they should not. I've been able to replicate the issue using a toy NF pipeline (see below). The problem only occurs when executing the pipeline on our shared FS (SMB3). The `file()` hash doesn't change - only the `ArrayBag` hash does.

### Program output 
Here's the relevant bits when running the toy NF pipeline with `-dump-hashes`

```
[Two (FOO)] cache hash: f85750fb9add10b280149fe04ce5bace; mode: STANDARD; entries:
465ed2e99af5b2877d4c36fd18167368 [nextflow.util.ArrayBag] [FileHolder(sourceObj:/imk/imk-bignas/developers/fmorency/Imeka/20180808-NF-HASH/Results_POC/work/85/3b3b31c5136fddf435db9beff7c292/one, storePath:/imk/imk-bignas/developers/fmorency/Imeka/20180808-NF-HASH/Results_POC/work/85/3b3b31c5136fddf435db9beff7c292/one, stageName:one)] 

[Three (FOO)] cache hash: 2c6cc3db7a91674429cf44b58d8ff4ec; mode: STANDARD; entries:
4208ad9f31d5e8becae51c3d8c31bb5d [nextflow.util.ArrayBag] [FileHolder(sourceObj:/imk/imk-bignas/developers/fmorency/Imeka/20180808-NF-HASH/Results_POC/work/85/3b3b31c5136fddf435db9beff7c292/one, storePath:/imk/imk-bignas/developers/fmorency/Imeka/20180808-NF-HASH/Results_POC/work/85/3b3b31c5136fddf435db9beff7c292/one, stageName:one)]
```

As you can see, the `ArrayBag` hashes are different for the exact same file.

The `ArrayBag` hash for `[...]/work/85/3b3b31c5136fddf435db9beff7c292/one` computed using `CacheHelper.hasher(bag).hash();` is `4208ad9f31d5e8becae51c3d8c31bb5d`

The process will be re-executed on the first `-resume` but not on the second one.

### Steps to reproduce the problem

main.nf
```
params.input = false

Channel.fromPath("$params.input")
    .map{ ["FOO", it] }
    .set{ch_one}

process One {
    input:
    set sid, file(one) from ch_one

    output:
    set sid, "one" into ch_one_for_two, ch_one_for_three

    script:
    """
    echo "one" > one
    sleep 15
    """
}

process Two {
    input:
    set sid, file(two) from ch_one_for_two

    output:
    set sid, "two" into ch_two_for_three

    script:
    """
    echo "two" > two
    sleep 15
    """
}

ch_one_for_three
    .join(ch_two_for_three)
    .set{ch_three}

process Three {
    input:
    set sid, file(one), file(two) from ch_three

    output:
    file "three"

    script:
    """
    echo "three" > three
    sleep 15
    """
}
```

nextflow.config
```
process {
    tag = { "$sid" }
    publishDir = {"./results/$sid/$task.process"}
    scratch = true
}
```

Create a `dummy` file and run with `nextflow run -dump-hashes main.nf --input dummy`

### Environment 

* Nextflow version: 0.30.2
* Java version: 1.8.0_171
* Operating system: Linux

**Comment by fmorency** [comment]
I should mention that the hashes seem to behave properly if I remove `scratch = true`

**Comment by pditommaso** [comment]
> The problem only occurs when executing the pipeline on our shared FS (SMB3)

This is a bad signal, therefore can't replicate the issue. 

> The file() hash doesn't change - only the ArrayBag hash does

Still missing evidence of this.

**Comment by pditommaso** [comment]
I've uploaded a snapshot with more detailed debugging. Run the above script as 

```
NXF_VER=0.31.2-SNAPSHOT nextflow -trace nextflow.util.CacheHelper -log a.log run -dump-hashes main.nf --input dummy
```

then 

```
NXF_VER=0.31.2-SNAPSHOT nextflow -trace nextflow.util.CacheHelper -log b.log run -dump-hashes -resume main.nf --input dummy
```


And upload both `a.log` and `b.log` here.

**Comment by fmorency** [comment]
Here's `a.log` and `b.log`. I also uploaded `c.log`and `d.log` which are essentially the same command as `b.log`. Jobs are being re-scheduled for `a`, `b` and `c` but not for `d`. You can see the progress of the issue and how NF handles the hashes.

[a.log](https://github.com/nextflow-io/nextflow/files/2275459/a.log)
[b.log](https://github.com/nextflow-io/nextflow/files/2275461/b.log)
[c.log](https://github.com/nextflow-io/nextflow/files/2275462/c.log)
[d.log](https://github.com/nextflow-io/nextflow/files/2275463/d.log)

**Comment by pditommaso** [comment]
Have you noticed that the file stamp is different in `a.log` and `b.log` 

a.log
```
Aug-09 14:16:35.240 [Actor Thread 6] TRACE nextflow.util.CacheHelper - Hash metadata file=/imk/imk-bignas/developers/fmorency/Imeka/20180808-NF-HASH/Results_POC_Snapshot/work/4f/49044071b122e837724e4b7cc435bb/one; size=4; lastmodified=1533838595205
```

b.log
```
Aug-09 14:18:35.878 [Actor Thread 7] TRACE nextflow.util.CacheHelper - Hash metadata file=/imk/imk-bignas/developers/fmorency/Imeka/20180808-NF-HASH/Results_POC_Snapshot/work/4f/49044071b122e837724e4b7cc435bb/one; size=4; lastmodified=1533838595206
```

**Comment by fmorency** [comment]
Yes I saw. Do you know what could cause this? It seems a (very) small difference.

**Comment by pditommaso** [comment]
Even a small diff results in a complete different hash number. 

No idea why it's happening, that info is read with the a plain [java api](https://github.com/nextflow-io/nextflow/blob/badcde7db76c1ed4ccf1194846c3065a035aa1dc/subprojects/nxf-commons/src/main/nextflow/util/CacheHelper.java#L187). I suggest there's some problem with the underlying file system. Have the two nextflow runs been launched in the same node ?

**Comment by fmorency** [comment]
Yes. I launched both using my desktop. I will try and ask around how this can happen on samba.

As a workaround, is there a way to disable the `lastmodified` bit from the hash calculation?

**Comment by pditommaso** [comment]
> is there a way to disable the lastmodified bit from the hash calculation?

No, but you can set `process.cache = 'deep'` to create the hash based on the file content instead of the metadata.

**Comment by fmorency** [comment]
The `process.cache = 'deep'` option works. I could `-resume` my task properly. As expected, the hash calculation takes (much) longer (around 35s per subject) which might become problematic on large studies. Is NF taking advantage of multiple threads to perform `-resume` and cache computation?

**Comment by fmorency** [comment]
Another thing that bugs me is the `lastmodified` difference which seems to be 1ms (right?). 

One the first run of the NF toy pipeline, the hash of the file `one` used in `process Three` is the right one. If I `-resume`, the `process Two` will get re-executed and the hash of the file `one` used in `process Two` will become the right one too. And so on... This is why if I keep `-resume` the pipeline will get to a point where it gets entirely cached and no process will get re-executed.

My questions are: when is the hash of the file calculated and put in the cache? Is it possible that the hash gets calculated "too soon"?

**Comment by pditommaso** [comment]
> Is NF taking advantage of multiple threads to perform -resume and cache computation?

Yes. 

> when is the hash of the file calculated and put in the cache?  

just before the task needs to be executed.

**Comment by fmorency** [comment]
Here's another interesting fact!

The POC doesn't re-execute any job if those have been submitted using the `slurm` executor. Before, I was using the local executor on my desktop + mounted SMB share. It seems like the lag introduced by using the `slurm` executor allows NF to behave correctly (at least when using the POC). I believe it's not always the case, such as when handling bigger files. I need to test this.

**Comment by fmorency** [comment]
Here's some more observations

- I can't reproduce the bug using local executor + shared FS if I remove the `sleep` from the processes. In that case, NF behaves correctly. In fact, only the `sleep` in `process Two` is required to reproduce the bug using local executor + shared FS.
- I can reproduce the bug using the original POC + local executor + shared FS on the head node. This is not an issue that only happens on my computer
- I can't reproduce the issue using a NFSv4 share
- If I modify the POC to add
```
process One {
   ...
   output:
   set sid, "one" into ch_one_for_two, ch_one_for_three, ch_one_for_four
   ...
}

process Four {
    input:
    set sid, file(four) from ch_one_for_four

    output:
    set sid, "four"

    script:
    """
    echo "four" > four
    """
}
```
then the hash for file `one` will be the same in `process Two` and `process Four` but different in `process Three`. According to my tests, the right one is the one from `process Three`.
- **The POC works using local scheduler + shared FS ONLY IF I place `afterScript 'sleep 1'` in `process One`**

I think at some point our shared FS has to modify/fix the file permissions to be compliant with the Samba share. However, I still don't understand what's going on. Do you have any idea what's going on between the moment `One` finishes and `Two` starts that could affect the hash?

**Comment by pditommaso** [comment]
Stil thinking this a glitch at network file system level, returning a different timestamp for the same file at different point in time. The only solution that I can see is to introduce a more loose hash calculation that rounds the file timestamp up to a second (for example) to try to bypass this problem. 

An estimation of much can vary the timestamp in such condition can be useful at this regard.

**Comment by fmorency** [comment]
I guess the variation will depends on the underlying network file system and the machine hosting it. Here, it seems to be around 1ms, maybe more if the host is under load. I added `afterScript 'sleep 1'` to our `nextflow.config` as a temporary workaround. I tested it extensively and it seems to always work.

You said on Gitter that you're aware of some network file system that can result in the same issue. Can you name a few for the sake of posterity? 

Do you want to implement an option/fix directly in NF? Or just document the issue and list the workaround?

**Comment by pditommaso** [comment]
> You said on Gitter that you're aware of some network file system that can result in the same issue. Can you name a few for the sake of posterity? 

I've seen this rarely happen on our NFS file system.

> Do you want to implement an option/fix directly in NF?

An option could be to introduce a new cache mode ie.`soft` that rounds file timestamp up to a second.

**Comment by pditommaso** [comment]
I've added a new *lenient* cache mode that just do not use file timestamp to calculate task unique id in the new snapshot `0.32.0u3-SNAPSHOT`. 

To enable it add in the `nextflow.config` file this line 

```
process.cache = 'lenient'
``` 

To run NF use this command: 

```
NXF_VER=0.32.0u3-SNAPSHOT nextflow run .. etc
```

**Comment by pditommaso** [comment]
reminder to myself, add the documentation.

**Comment by pditommaso** [comment]
Actually this was not merged in the master

**Comment by pditommaso** [comment]
Released with version `18.10.1`.
